### ISSUE TITLE
Possible bad hash distribution for strings

### ISSUE THREAD

**Issue by theseion** [pharo-project/pharo-vm#538]
@guillep 

As discovered in https://github.com/theseion/Fuel/issues/269, Pharo 11 (or rather its VM) appears to compute identity hashes for strings differently than before. I'm not really sure that this is an issue, but I still wanted to report what I've found.

```smalltalk
(1 to: 10000) collect: [ :i | (i asString basicIdentityHash) \\ 4096 ]
```

The above will generate 4096 unique values in Pharo 10 but only 2048 in Pharo 11. The set in Pharo 11 does not include even numbers.
This suggests to me that the space of identity hashes for strings _may_ be half as big as it should be, which could double hash collisions and reduce performance of hashed collections.

**Comment by guillep** [comment]
Yesterday i took a look at how does it scale by giving fuel a number of strings obtained by tokenising lorem ipsum strings of one MB bigger each time.

I'm far from my machine, i'll share some plots next week.

What i can tell is that up to ~14MB it was faster usign the system identity dictionary scaling somehow linearly in the order of a couple of seconds. After 14MB, it jumps from a couple of seconds to a couple lf minutes.

Using the fuel large dictionary, scaling seems linear up to the points I saw, but slower than the normal identity dictionary for small(ish) sets.

Now, in my case i happen to have lots of equal but unique strings.

Strings should not have a special definition of identity hash, so this issue should be general to any large set of objects of the same class, right ? 

If that is the case, maybe we need to study the distribution of the identity hash (that is attributed by the VM) and evaluate a separate hash table implementation that better handles conflicts for these cases of large sets ? Otherwise, we will probably end up with lots of linear searches that kind of kill the purpose of the library :)

**Comment by theseion** [comment]
> Using the fuel large dictionary, scaling seems linear up to the points I saw, but slower than the normal identity dictionary for small(ish) sets.

Makes sense. That's what they were designed to do. But I'm surprised how slow `FLLargeIdentityDictionary is`. I suspect the reason is that too many scans happen.

> Strings should not have a special definition of identity hash, so this issue should be general to any large set of objects of the same class, right ?

Looks like it. Using `Object new` instead of `i asString` also yields 2048 unique hashes.

**Comment by guillep** [comment]
I'll leave here the (small) analysis I did before vacations. I will say before that I don't think this is a VM issue. The VM has to ensure that hashes are unique and fastly allocated... I'd say that we need specific hash tables for lots of elements based on identity hashes.

That being said, what I did was to compare the scalability of fuel serialization of large arrays of strings computed in the following way:

```
initial := 10.
step := 1000000.
previous := initial.
times := OrderedCollection new.
14 timesRepeat: [
	li := String loremIpsum: previous.
	li := li lines collect: [ :e | e substrings ].
	times add: [FLSerializer 
		serialize: li
		toFileNamed: previous asString, '-loremipsum.fuelv2'.]
			timeToRun.
	previous := previous + step
].
```

I did three versions of this:
 - the baseline (orange below) serialises the array using STON
 - then default fuel with `FLLargeIdentityDictionary` (dark blue)
 - then fuel using a standard `IdentityDictionary` (light blue)

<img width="587" alt="imagen" src="https://user-images.githubusercontent.com/708322/220942407-11ab825b-ef1c-4ab1-808f-f59e5dbae9bd.png">

What we can see is that a standard identity dictionary scales better up to a point, then performance suffers miserably.
That point happened to be for me around fuel files of ~16MB.
